pkg/common/cns-lib/vsphere/virtualcenter.go

```diff
@@ -228,11 +231,23 @@ func (vc *VirtualCenter) connect(ctx context.Context, requestNewSession bool) er
 		sessionMgr := session.NewManager(vc.Client.Client)
 		// SessionMgr.UserSession(ctx) retrieves and returns the SessionManager's CurrentSession field
 		// Nil is returned if the session is not authenticated or timed out.
-		if userSession, err := sessionMgr.UserSession(ctx); err != nil {
+		var userSession *types.UserSession
+		if userSession, err = sessionMgr.UserSession(ctx); err != nil {
 			log.Errorf("failed to obtain user session with err: %v", err)
 			return err
-		} else if userSession != nil {
-			return nil
+		}
+		if userSession != nil {
```

Comment:
Thinking a bit more, I'm not sure how is this fixing the problem. In the earlier code, if `sessionMgr.UserSession` returned a nil `userSession`, the code would go ahead and create a new VC client that uses the VC creds for login. What is wrong with the earlier code and why do we need to check if the current session is active? Are you suspecting that `sessionMgr.UserSession` returns a session that may not be active and possibly NotAuthenticated?

Reply:
In the earlier code, we were not checking if the returned session is active. This change is helping to validate if the returned session is still valid and active using `sessionMgr.SessionIsActive(ctx)`, if not, we are requesting a new session.

customer observed following issue 

> E1117 14:40:14.340336       1 controller.go:908] error syncing claim "1d120f42-6b31-4b39-8029-a7a266469344": failed to provision volume with StorageClass "csi-sc": rpc error: code = Internal desc = Failed to create volume. Error: ServerFaultCode: NotAuthenticated

means sessionMgr.UserSession is returning the session, with no error, but that session is NotAuthenticated.

Reply:
@dougm do you know when such a situation can happen? 
Do you think additional check using `sessionMgr.SessionIsActive(ctx)` will help improve here?

Reply:
I don't know that adding SessionIsActive will improve, since the UserSession check is essentially the same.  Ideally the problem could be reproduced to find the proper solution. There is a way to terminate sessions which may help in reproducing: https://github.com/vmware/govmomi/blob/master/govc/USAGE.md#sessionrm
You can use `session.ls` and grep for the csi driver user-agent to find the right session to terminate.

Reply:
@dougm I have tried out session.rm, but unable to reproduce the NotAuthenticated case.